File: SIMPLEAPI_Demo/SIMPLEAPI_Demo/Impresion/Core/Classes/PrinterService.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Printing;
using System.Linq;
using System.Windows.Forms;

namespace SIMPLEAPI_Demo.Impresion.Core.Classes
{
    public class PrinterService
    {
        List<Font> Fonts { get; set; }
        List<SolidBrush> BrushesPallete { get; set; }
        PrinterWork Work { get; set; }

        enum WrapResolve
        {
            Truncate = 0,
            Wrap = 1
        }

        public PrinterService(PrinterWork work)
        {
            this.Work = work;
            BrushesPallete = new List<SolidBrush>();
            Fonts = new List<Font>();
        }

        /// <summary>
        /// Printer previously generated PrinterWork
        /// </summary>
        /// <param name="documentName">Nombre del documento a imprimir.</param>
        /// <param name="printerName"></param>
        /// <param name="debugMode"></param>
        public void Print(string documentName, string printerName, bool debugMode = false)
        {
            if (string.IsNullOrEmpty(documentName))
                throw new ArgumentException("El parámetro 'documentName' no puede ser 'null' o estar vacío", documentName);

            if (string.IsNullOrEmpty(printerName))
                throw new ArgumentException("El parámetro 'printerName' no puede ser 'null' o estar vacío", printerName);

            if (Work == null)
                throw new ArgumentException("No se puede imprimir si la propiedad 'Work' es null", printerName);

            if (Work.Rows.Count == 0)
                return;

            PrintDocument pdoc = new PrintDocument();
            pdoc.DocumentName = documentName;
            if (!debugMode)
                pdoc.PrintController = new StandardPrintController(); // hides status dialog popup

            pdoc.DefaultPageSettings.PaperSize = new PaperSize("Custom", 100, 200);
            pdoc.DefaultPageSettings.PaperSize.Height = Work.Rows.Count * 24; // Aprox
            pdoc.DefaultPageSettings.PaperSize.Width = 300;
            pdoc.PrintPage += new PrintPageEventHandler(PrintEvent);

            if (!debugMode)
            {
                PrinterSettings ps = new PrinterSettings();
                ps.PrinterName = printerName;
                pdoc.PrinterSettings = ps;
                pdoc.Print();
            }
            else
            {
                PrintPreviewDialog printPrvDlg = new PrintPreviewDialog();
                printPrvDlg.Document = pdoc;
                printPrvDlg.Width = 1200;
                printPrvDlg.Height = 800;
                printPrvDlg.ShowDialog();
            }
        }
        private void PrintEvent(object sender, PrintPageEventArgs e)
        {
            Graphics graphics = e.Graphics;
            SolidBrush brush;
            Font font;

            int currentY = 0;

            foreach (var row in Work.Rows)
            {
                brush = GetBrush(row.Color);
                font = GetFont(row.FontSize, row.IsBold);

                // Kind of excesive but will throw an exception if sums of width percentages of all row colums is not 1
                row.ValidateColumnsWidth(Work.IsTestWork);

                foreach (var line in row.ToLines())
                {
                    if (line.StartsWith("<img>") && line.EndsWith("</img>"))
                    {
                        var col = row.Columns.First() as PrinterImageColumn;

                        graphics.DrawImage(col.Image, 0, currentY, 290, 120);
                        currentY += 120;
                        break;
                    }
                    else
                    {
                        graphics.DrawString(line, font, brush, 0, currentY);
                        currentY += (int)font.GetHeight();
                    }
                }
            }
        }

        #region Helpers

        private Font GetFont(float size, bool isBold)
        {
            var font = Fonts.FirstOrDefault(x => x.Size == size && (isBold ? x.Style == FontStyle.Bold : x.Style == FontStyle.Regular));
            if (font == null)
            {
                font = new Font(FontInfo.FontName, size, (isBold ? FontStyle.Bold : FontStyle.Regular));
                Fonts.Add(font);
            }
            return font;
        }

        private SolidBrush GetBrush(Color color)
        {
            var brush = BrushesPallete.FirstOrDefault(x => x.Color == color);
            if (brush == null)
            {
                brush = new SolidBrush(color);
                BrushesPallete.Add(new SolidBrush(color));
            }
            return brush;
        }

        #endregion
    }
}

File: SIMPLEAPI_Demo/SIMPLEAPI_Demo_VB/Impresion/Core/Classes/PrinterService.vb


Imports System.Drawing

Imports System.Drawing.Printing

Public Class PrinterService
    Private Property Fonts As List(Of Font)
    Private Property BrushesPallete As List(Of SolidBrush)
    Private Property Work As printerWork

    Enum WrapResolve
        Truncate = 0
        Wrap = 1
    End Enum

    Public Sub New(ByVal work As printerWork)
        Me.Work = work
        BrushesPallete = New List(Of SolidBrush)()
        Fonts = New List(Of Font)()
    End Sub

    '''<summary>
    '''Printer previously generated PrinterWork
    ''' </summary>
    '''<param name="documentName">Nombre del documento a imprimir.</param>
    '''<param name="printerName"></param>
    '''<param name="debugMode"></param>

    Public Sub Print(ByVal documentName As String, ByVal printerName As String, ByVal Optional debugMode As Boolean = False)
        If String.IsNullOrEmpty(documentName) Then Throw New ArgumentException("El parámetro 'documentName' no puede ser 'null' o estar vacío", documentName)
        If String.IsNullOrEmpty(printerName) Then Throw New ArgumentException("El parámetro 'printerName' no puede ser 'null' o estar vacío", printerName)
        If Work Is Nothing Then Throw New ArgumentException("No se puede imprimir si la propiedad 'Work' es null", printerName)
        If Work.Rows.Count = 0 Then Return
        Dim pdoc As PrintDocument = New PrintDocument()
        pdoc.DocumentName = documentName
        If Not debugMode Then pdoc.PrintController = New StandardPrintController()
        pdoc.DefaultPageSettings.PaperSize = New PaperSize("Custom", 100, 200)
        pdoc.DefaultPageSettings.PaperSize.Height = Work.Rows.Count * 24
        pdoc.DefaultPageSettings.PaperSize.Width = 300
        AddHandler pdoc.PrintPage, New PrintPageEventHandler(AddressOf PrintEvent)

        If Not debugMode Then
            Dim ps As PrinterSettings = New PrinterSettings()
            ps.PrinterName = printerName
            pdoc.PrinterSettings = ps
            pdoc.Print()
        Else
            Dim printPrvDlg As PrintPreviewDialog = New PrintPreviewDialog()
            printPrvDlg.Document = pdoc
            printPrvDlg.Width = 1200
            printPrvDlg.Height = 800
            printPrvDlg.ShowDialog()
        End If
    End Sub
    Private Sub PrintEvent(ByVal sender As Object, ByVal e As PrintPageEventArgs)
        Dim graphics As Graphics = e.Graphics
        Dim brush As SolidBrush
        Dim font As Font
        Dim currentY As Integer = 0

        For Each row In Work.Rows
            brush = GetBrush(row.Color)
            font = GetFont(row.FontSize, row.IsBold)
            row.ValidateColumnsWidth(Work.IsTestWork)

            For Each line In row.ToLines()

                If line.StartsWith("<img>") AndAlso line.EndsWith("</img>") Then
                    Dim col = TryCast(row.Columns.First(), PrinterImageColumn)
                    graphics.DrawImage(col.Image, 0, currentY, 290, 120)
                    currentY += 120
                    Exit For
                Else
                    graphics.DrawString(line, font, brush, 0, currentY)
                    currentY += CInt(font.GetHeight())
                End If
            Next
        Next
    End Sub

#Region "Helpers"
    Private Function GetFont(ByVal size As Single, ByVal isBold As Boolean) As Font
        Dim font = Fonts.FirstOrDefault(Function(x) x.Size = size AndAlso (If(isBold, x.Style = FontStyle.Bold, x.Style = FontStyle.Regular)))

        If font Is Nothing Then
            font = New Font(FontInfo.FontName, size, (If(isBold, FontStyle.Bold, FontStyle.Regular)))
            Fonts.Add(font)
        End If

        Return font
    End Function

    Private Function GetBrush(ByVal color As Color) As SolidBrush
        Dim brush = BrushesPallete.FirstOrDefault(Function(x) x.Color = color)

        If brush Is Nothing Then
            brush = New SolidBrush(color)
            BrushesPallete.Add(New SolidBrush(color))
        End If

        Return brush
    End Function
#End Region
End Class
